### PR TITLE
Add CfCT version for deployment

### PR DESCRIPTION
**Pull request by str3tch**
*Issue #, if available:*

*Description of changes:*
Added a version which is compatible with CfCT.. see readme for details

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          - ServerSideEncryptionByDefault:
+              KMSMasterKeyID: !Ref KMSMasterKeyID
+              SSEAlgorithm: !Ref SSEAlgorithm
+    Metadata:
+      cfn_nag:
+        rules_to_suppress:
+          - id: W35
+            reason: "Supress bucket access logging requirement"
+
+  BucketPolicy:
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:70]: can you add 1 extra policy to enforce secure SSL as shown in [c9a6e3e](https://github.com/aws-samples/aws-ct-vpc-flowlogs/pull/6/commits/a52873a911d390010d47c761233df3fe4228dfd9)

**Review by wellsiau-aws — CHANGES_REQUESTED** [review]
thanks for helping with this, I have small ask to tighten the bucket policy for secure transport, please have a look.

File: cfct/manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    deployment_targets:
+      accounts:
+        - Management
+    regions:
+      - ap-southeast-2
+
+  - name: VPCFlowLogRoleInSpoke
+    resource_file: templates/vpc_flowlog_spoke.template
+    parameters:
+      - parameter_key: "OrganizationId"
```
Comment (wellsiau-aws) [inline cfct/manifest.yaml:57]: this param does not exist in the template
error : An error occurred (ValidationError) when calling the CreateStackSet operation: Parameters: [ControlTowerMasterRegion, OrganizationId] do not exist in the template

File: cfct/manifest.yaml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        parameter_value: "$[alfred_ssm_/org/core/OrganizationID]"
+
+      - parameter_key: "EventBusDestinationAccount" # Network hub account
+        parameter_value: "<REPLACE ME>"
+
+      - parameter_key: "EventBusName"
+        parameter_value: "FlowLog-EventBus"
+
+      - parameter_key: "ControlTowerMasterRegion"
```
Comment (wellsiau-aws) [inline cfct/manifest.yaml:66]: same as previous, this param also does not exist in the template

File: cfct/manifest.yaml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        parameter_value: "24"
+
+      - parameter_key: "BaselineCloudTrailStackSetName"
+        parameter_value: "AWSControlTowerBP-BASELINE-CLOUDTRAIL"  # do not change this, used to get all accounts managed by Control Tower.
+
+      - parameter_key: "BaselineCloudTrailStackArn"
+        parameter_value: "<REPLACE ME>"  # find this from mgmt account in cloudformation stacksets
+
+      - parameter_key: "S3LambdaBucket"
+        parameter_value: "$[alfred_ssm_/org/primary/storagebucket]"
```
Comment (wellsiau-aws) [inline cfct/manifest.yaml:109]: this SSM param does not exist by default, I suggest to make call out in the README.md

File: cfct/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+      - name: /org/sharedservice/networking/vpcflowlogsbucket
+        value: $[output_oBucketName]
+    regions:
+      - ap-southeast-2
+
+  - name: VPCFlowLogsMgmtRole
+    resource_file: templates/vpc_flowlog_mgmt.template
+    parameters:
+      - parameter_key: "OrganizationId"
+        parameter_value: "$[alfred_ssm_/org/core/OrganizationID]"
```
Comment (wellsiau-aws) [inline cfct/manifest.yaml:40]: this SSM param does not exist by default, I suggest to make call out in the README

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        Version: 2012-10-17
+        Statement:
+          - Sid: AWSLogDeliveryWrite
+            Effect: Allow
+            Principal:
+              Service: 'delivery.logs.amazonaws.com'
+            Action:
+              - 's3:PutObject'
+            Resource:
+              - !Sub "control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}/*"
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:89]: resource name invalid, it should include: `arn:aws:s3:::control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}/*`

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+              StringEquals:
+                s3:x-amz-acl: bucket-owner-full-control
+          - Sid: AWSLogDeliveryAclCheck
+            Effect: Allow
+            Principal:
+              Service: 'delivery.logs.amazonaws.com'
+            Action:
+              - 's3:GetBucketAcl'
+            Resource:
+              - !Sub "control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}"
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:100]: this should be `arn:aws:s3:::control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}`

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+          - Sid: AllowOrganizationRead
+            Effect: Allow
+            Principal: "*"
+            Action: 
+              - s3:GetObject
+            Resource:
+              - !Sub "control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}/*"
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:107]: this should be `arn:aws:s3:::control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}/*`

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    - 'AES256'
+
+Resources:
+
+  S3BucketFlowLoggingSSE:
+    Type: AWS::S3::Bucket
+    Condition: UseAES256
+    DeletionPolicy: Retain
+    UpdateReplacePolicy: Retain
+    Properties:
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:40]: bucket needs to have versioning, ref [cb9424b](https://github.com/aws-samples/aws-ct-vpc-flowlogs/pull/7/commits/cb9424b557ddad6d9a33a284f737868b5c88ebdf)

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        rules_to_suppress:
+          - id: W35
+            reason: "Supress bucket access logging requirement"
+
+  S3BucketFlowLoggingKMS:
+    Type: AWS::S3::Bucket
+    Condition: UseKMS
+    DeletionPolicy: Retain
+    UpdateReplacePolicy: Retain
+    Properties:
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:57]: needs bucket versioning, ref [cb9424b](https://github.com/aws-samples/aws-ct-vpc-flowlogs/pull/7/commits/cb9424b557ddad6d9a33a284f737868b5c88ebdf)

File: cfct/manifest.yaml [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+      - parameter_key: "EventBusDestinationAccount" # Typiclly this will be the NetworkHub account
+        parameter_value: "<REPLACE ME>"
+
+      - parameter_key: "BaselineCloudTrailStackArn" # Copy this this from mgmt account in cloudformation stacksets for AWSControlTowerBP-BASELINE-CLOUDTRAIL
+        parameter_value: "<REPLACE ME>"
+    deploy_method: stack_set
+    deployment_targets:
+      accounts:
+        - Management
```
Comment (wellsiau-aws) [inline cfct/manifest.yaml:50]: should we make call out in the README that `AWSControlTowerExecution` must exist on the Management account?

File: cfct/template/vpc_flowlog_spoke.template [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        }
+      State: ENABLED
+      Targets:
+        - Arn: !Sub arn:aws:events:${AWS::Region}:${EventBusDestinationAccount}:event-bus/${EventBusName}
+          Id: "TagCreateUpdateTrigger"
+          RoleArn: !GetAtt FlowLogTagSpokeRuleDeliveryRole.Arn
+
+  FlowLogTagSpokeRuleDeliveryRole:
+    Type: AWS::IAM::Role
+    Properties:
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_spoke.template:64]: Missing role name. The Lambda function at the Hub account needs to know which role it should assume on each spoke accounts. Perhaps we could avoid hard coded role name and pass it as parameter?

File: cfct/manifest.yaml (hunk 2 above)
Comment (wellsiau-aws) [reply cfct/manifest.yaml:57]: actually ignore this - the real cause is due to missing param `OrganizationId` on the template itself.
this param is required in order to create logical resource `FlowLogHubAssumeRole` which also missing from the `vpc_flowlog_spoke.template`  

Ref to [FlowLogHubAssumeRole](https://github.com/aws-samples/aws-ct-vpc-flowlogs/blob/main/template/ct_vpc_flowlog_stackset.yml#L379)

Comment (str3tch) [reply cfct/manifest.yaml:57]: yep I missed out my latest change which had the role in the spoke template.. latest commit will fix this

File: cfct/manifest.yaml (hunk 4 above)
Comment (str3tch) [reply cfct/manifest.yaml:109]: removed use of alfred, for ease of deployment

File: cfct/template/vpc_flowlog_logging_bucket.template (hunk 9 above)
Comment (str3tch) [reply cfct/template/vpc_flowlog_logging_bucket.template:40]: done

File: cfct/template/vpc_flowlog_logging_bucket.template (hunk 10 above)
Comment (str3tch) [reply cfct/template/vpc_flowlog_logging_bucket.template:57]: done

File: cfct/manifest.yaml (hunk 5 above)
Comment (str3tch) [reply cfct/manifest.yaml]: I added the management_prep.template which adds this param in, along with the AWSControlTowerExecution role

File: cfct/manifest.yaml (hunk 3 above)
Comment (str3tch) [reply cfct/manifest.yaml]: fixed

File: cfct/manifest.yaml (hunk 4 above)
Comment (str3tch) [reply cfct/manifest.yaml:120]: Now available via management_prep.template, which I call out in the readme as a pre-req

File: cfct/template/vpc_flowlog_logging_bucket.template (hunk 1 above)
Comment (str3tch) [reply cfct/template/vpc_flowlog_logging_bucket.template:85]: done

File: code/ct_flowlog_activator.py [hunk 13]
```diff
@@ -335,81 +381,161 @@ def list_stack_instance_by_region(target_session, stack_set_name, region):
...
+
     except Exception as e:
-        LOGGER.error("List Stack Instance error: %s" % e)
+        LOGGER.error("List Stack Instance error: %s" % e, exc_info=True)
         return []
 
-def lambda_handler(event, context):    
+
+def lambda_handler(event, context):
     LOGGER.info('Lambda Handler - Start')
```
Comment (str3tch) [inline code/ct_flowlog_activator.py:393]: @wellsiau-aws Did you check this diff? keen to have your eyes on this bit of code change

Comment (wellsiau-aws) [reply code/ct_flowlog_activator.py:393]: honestly, I missed it and then later realized it when I ran my test, I'll look at it again

File: cfct/template/vpc_flowlog_logging_bucket.template [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                s3:x-amz-acl: bucket-owner-full-control
+          - Sid: AWSLogDeliveryAclCheck
+            Effect: Allow
+            Principal:
+              Service: 'delivery.logs.amazonaws.com'
+            Action:
+              - 's3:GetBucketAcl'
+            Resource:
+              - !Sub "control-tower-vpcflowlogs-${AWS::AccountId}-${AWS::Region}"
+          - Sid: AllowOrganizationRead
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_logging_bucket.template:117]: IMHO, this should be optional, customer should define how they want to allow read access to this bucket on case-by-case basis.

File: cfct/template/vpc_flowlog_automation_in_hub.template [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+            Resource: 
+              - !Join ['', ['arn:aws:iam::', '*', ':role/', !FindInMap [LambdaVariable,Role, Spoke]]]
+              - !Join ['', ['arn:aws:iam::', !Ref MgmtAccountId, ':role/', !FindInMap [LambdaVariable,Role, Hub]]]
+            Condition:
+              StringEquals:
+                "sts:ExternalId": !Ref OrganizationId
+          - Effect: Allow
+            Action:
+              - lambda:InvokeFunction
+            Resource: !Sub 'arn:aws:lambda:*:${AWS::AccountId}:function:${AWS::AccountId}-FlowLogActivator'
```
Comment (wellsiau-aws) [inline cfct/template/vpc_flowlog_automation_in_hub.template:197]: this should be
'arn:aws:lambda:*:${AWS::AccountId}:function:FlowLogActivator'

File: cfct/template/vpc_flowlog_logging_bucket.template (hunk 14 above)
Comment (str3tch) [reply cfct/template/vpc_flowlog_logging_bucket.template]: I have added a comment that they could/should change

**Review by wellsiau-aws — APPROVED** [review]
LGTM

File: code/ct_flowlog_activator.py (hunk 13 above)
Comment (wellsiau-aws) [reply code/ct_flowlog_activator.py:393]: found issue with https://github.com/aws-samples/aws-ct-vpc-flowlogs/issues/8, which I will address as separate PR.

**Pull request merged**